LOPS-741 - PantheonSecrets Documentation (#8975)
pantheon-systems/documentation: source/content/guides/secrets/02-basic-concepts.md

---
title: Pantheon Secrets Guide
subtitle: Basic Concepts
description: Gaining familiarity with some concepts about Pantheon Secrets will help you make the most of this feature.
contributors: [stovak]
contenttype: [guide]
innav: [true]
categories: [secrets]
cms: [drupal, wordpress]
audience: [development]
product: [secrets]
integration: [--]
tags: [reference, cli, local, terminus, workflow]
permalink: docs/guides/secrets/basic-concepts
reviewed: "2024-05-01"
---
<dl>  

<dt>Secret</dt>

<dd>

A key-value pair that should not be exposed to the general public, typically something like a password, API key, or other sensitive information that should not be added to version control.

</dd>

<dt>Secret type <sup>1</sup></dt>

<dd>

  This is a field on the secret record. It defines the usage for this secret and how it is consumed. Current types are:

>   * `runtime`: this secret will be used to retrieve it in application runtime using API calls to the secret service. This is the recommended type to set if you want your secret to be exposed to the application runtime.

Reviewer: I fixed it on ~Friday and I don't remember what the problem was anymore.  It honestly might have been white space.

```mermaid
classDiagram
OrganizationSecretAPIPassword --> SiteSecretAPIPassword 
SiteSecretAPIPassword  --> IntegratedComposerAPIPassword : no overrides
OrganizationSecretAPIPassword : string name apipassword
OrganizationSecretAPIPassword : string value ball00n
SiteSecretAPIPassword : Inherits value from Org 
SiteSecretAPIPassword : No Overrides
IntegratedComposerAPIPassword: value ball00n

OrganizationSecretOverrideExample --> SiteSecretOverrideExample
SiteSecretOverrideExample --> SiteSecretOverrideExampleDev : default value
SiteSecretOverrideExample --> SiteSecretOverrideExampleTest : env override value
SiteSecretOverrideExample --> SiteSecretOverrideExampleLive : env override value
OrganizationSecretOverrideExample : string name apipassword
OrganizationSecretOverrideExample : string value ball00n
SiteSecretOverrideExample : Inherits value from Org 
SiteSecretOverrideExample : No Site Overrides
SiteSecretOverrideExampleDev: value ball00n
SiteSecretOverrideExampleDev: defaultValue()
SiteSecretOverrideExampleTest: value ball00n2
SiteSecretOverrideExampleTest: overridden()
SiteSecretOverrideExampleLive: value ball00n3
SiteSecretOverrideExampleLive: overridden()

> ```

Reviewer: Converted to an image, but the docs site seems to be making the background smeary.

### Site-owned secrets
Site-owned secrets are available to the site and all of its environments. A common use-case is Github tokens that a site's composer build can use to access private repos referenced in the composer file.

### Environment override
Environment overrides provide overrides to a secret value for a specific environment. A common use case for this are API keys that are different in production and non-production environments.

## Value Resolution

1. Organization values have the lowest priority. They form the base value that is used when there is no more specific value provided for the site or environment.

3. Site values will replace the organization values when present. To return the secret to it's organization value, simply delete the site value.

4. Environmental overrides have the highest priority. If the override exists, it will become the value provided to the calling function.

> ![Secrets Relationships](../../../images/guides/secrets/secrets-relationships.png)

Reviewer 2: > Maybe we could replace the diagram with something simpler, similar to this diagram: ![Screenshot 2024-07-23 at 11 22 10 AM](https://private-user-images.githubusercontent.com/612191/351445124-cb81504a-37e5-4a32-bb1d-7abb36579138.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzNzAxOTIsIm5iZiI6MTcyMjM2OTg5MiwicGF0aCI6Ii82MTIxOTEvMzUxNDQ1MTI0LWNiODE1MDRhLTM3ZTUtNGEzMi1iYjFkLTdhYmIzNjU3OTEzOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDczMFQyMDA0NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZjhiNzJmNzliN2Q4ODk5NWMwYWIxMzc3MzUwMzhiY2FkOGY4MzhhZDZlYjIwYzQxOWJhZDM0YzBlZGU2ZTdiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.VtMefR1wGVVFZXTixdGxoXPZyM3KMx8SsLrPNSwhJqE) Perhaps this goes in the "Value Resolution" section.

[Requesting assist from @mel-miller](https://github.com/pantheon-systems/documentation/pull/8975/files/525091d85a05efa3733aa2be63ad8b1548b9f16e#r1697458982)

---
<dl>  

<dt>Secret</dt>

<dd>

A key-value pair that should not be exposed to the general public, typically something like a password, API key, or other sensitive information that you should probably not add to version control.

</dd>

<dt>Secret type [^1]</dt>

<dd>

  This is a field on the secret record. It defines the usage for this secret and how it is consumed. Current types are:

  * runtime: this secret will be used to retrieve it in application runtime using API calls to the secret service. This is the recommended type to set if you want your secret to be exposed to the application runtime.

  * env: this secret will be used to set environment variables in the application runtime. This type is currently only supported for Integrated Composer builds and not being exposed as environment variables to the application runtime.

  * composer: this secret type is used for composer authentication to private packages and it is the recommended method for doing so.

</dd>

<dt>Secret Scope</dt>

<dd>
A secret's scope is the answer to the question "Where is the secret's value available?". Once, set a secret's scope cannot be changed. The secret must be deleted and recreated.

>   * `runtime`: this secret will be used to retrieve it in application runtime using API calls to the secret service. This is the recommended type to set if you want your secret to be exposed to the application runtime.

Reviewer 3: cut-copy-pasta error: this section lists the secret types again, from the section above, instead of describing the available scopes.

Reviewer 3: https://github.com/pantheon-systems/terminus-secrets-manager-plugin?tab=readme-ov-file#secret-scope

Reviewer: I believe I resolved the `code` formatting so both types and scopes use `code`.  It looks okay in my local environment but the PR previews don't seem to be building anymore.

>   * `runtime`: this secret will be used to retrieve it in application runtime using API calls to the secret service. This is the recommended type to set if you want your secret to be exposed to the application runtime.

Reviewer: >   * `runtime`: This secret type can be retreived directly from your application code.  This is the recommended type if you want your application to be able to use the secret while it's operating.

>   * `env`: this secret will be used to set environment variables in the application runtime. This type is currently only supported for Integrated Composer builds and not being exposed as environment variables to the application runtime.

Reviewer: > * `composer`: This secret type is specifically used for authentication when pulling Composer packages from private repositories.  This is the recommended method for installing private composer packages.

  * `composer`: this secret type is used for composer  authentication to private packages and it is the recommended method for doing so.

</dd>

<dt>Owning Entity</dt>

<dd>

> * **Organization-owned secrets** -  Organization-owned secrets are available to every site and environment that are associated with the owning organization. A common use-cases is for a CI system and infrastructure that's shared among all sites in an organization.

Reviewer 3: We should explicitly note that secrets from supporting organizations are **not** shared with sites that they support.

Author: https://github.com/pantheon-systems/documentation/pull/8975/commits/fb9a303efaf79965e1ba2878ae846dc837d4c09b

* **Site-owned secrets** -  Site-owned secrets are available to the site and all of it's environments. A common use-case is Github tokens that a site's composer build can use to access private repos referenced in the composer file.

> * **Environment override**[^d]  - Environment overrides provide overrides to a secret value for a specific environment. A common use case for this are API keys that are different in production and non-production environments.

Author: I just put a "^d" there and my markdown editor supplies the number on render. Footnote didn't get moved over.

</dd>

<dt>Value Resolution</dt>

<dd>

> 1. Organization values are resolved first. They form the base value for the key-pair's value resolution.

Reviewer 3: Organization-owned secrets can also have environment overrides, but only dev, test and live are supported.

Reviewer: oh, and that's different from site level?  man that's confusing...

Author: which is why I made the graph I just added.

Reviewer: Unresolving because the mermaid diagram doesn't render on the docs site.  Needs to be converted to an image.

---
title: Pantheon Secrets Guide
subtitle: Basic Concepts
description: Gaining familiarity with some concepts about Pantheon Secrets will help you make the most of this feature.
contributors: [stovak]
contenttype: [guide]
innav: [true]
categories: [secrets]
cms: [drupal, wordpress]
audience: [development]
product: [secrets]
integration: [--]
tags: [reference, cli, local, terminus, workflow]
permalink: docs/guides/secrets/basic-concepts
reviewed: "2024-05-01"
---
<dl>

<dt>Secret</dt>

<dd>

> A key-value pair that should not be exposed to the general public, typically something like a password, API key, or other sensitive information that should not be added to version control.

Reviewer 3: ```suggestion
A secret is a key-value pair that should not be exposed to the general public, typically something like a password, API key, or other sensitive information that should not be added to version control.
```

</dd>

<dt>Secret type<sup>1</sup></dt>

<dd>
This represents how the secret is used.  A secret can only have one type.

Current types are:

>   * `runtime`: This secret type can be retreived directly from your application code.  This is the recommended type if you want your application to be able to use the secret while it's operating.

Reviewer 3: Maybe name-drop `pantheon_get_secret()` here, so that folks who are skimming the documentation get the information they need sooner (and do not decide to just use type env with "getenv()")

Reviewer: Added

>   * `env`: This type is used to set environment variables. Environment variables are currently only supported for Integrated Composer builds; setting environment variables on the application server is unsupported.

Reviewer 3: Once LOPS-1977 is done, we'll need to update this to describe how users can opt-in to use secrets of type "env" from the "web" scope. Point out that secrets of type env should not be considered to be encrypted at rest, and that  environment variables are easy (easier?) to accidentally expose.

  * `composer`: This secret type is specifically used for authentication when pulling Composer packages from private repositories.  This is the recommended method for installing private composer packages.

</dd>

<dt>Secret Scope</dt>

<dd>

> A secret's scope is the answer to the question "Where is the secret's value available?". Once set, a secret's scope cannot be changed. The secret must be deleted and recreated.

Reviewer 3: ```suggestion
A secret's scope is the answer to the question "Where is the secret's value available?". Once set, a secret's scope cannot be changed. The secret must be deleted and recreated to change its scope.
```

>   * `ic`: This secret will be readable by the Integrated Composer runtime. You should use this scope to get access to your private repositories.

Reviewer 3: ```suggestion
  * `ic`: This secret will be readable during Integrated Composer builds. You should use this scope to get access to your private repositories.
```

  * `web`: this secret will be readable by the application runtime.

>   * `user`: this secret will be readable by the user. This scope should be set if you need to retrieve the secret value at a later stage.

Reviewer 3: ```suggestion
  * `user`: this secret will be readable by the user. This scope should be set if you want to see the value of your secret displayed when listing site secrets with Terminus. The value for secrets without the the user scope is redacted in the Terminus secrets list.
```

</dd>

<dt>Owning Entity</dt>

<dd>

* **Organization-owned secrets** - Organization-owned secrets are available to every site and environment that are associated with the owning organization. A common use-cases is for a CI system and infrastructure that's shared among all sites in an organization. Note that secrets from "Supporting" Organizations are explicitly ***not shared*** with the sites they support. Sites receive secret key/value pairs from their Primary Organization only.

* **Site-owned secrets** - Site-owned secrets are available to the site and all of its environments. A common use-case is Github tokens that a site's composer build can use to access private repos referenced in the composer file.

* **Environment override** - Environment overrides provide overrides to a secret value for a specific environment. A common use case for this are API keys that are different in production and non-production environments.

> ![Secrets Relationships](../../../images/guides/secrets/secrets-relationships.png)

Reviewer 3: Maybe we could replace the diagram with something simpler, similar to this diagram:
![Screenshot 2024-07-23 at 11 22 10 AM](https://github.com/user-attachments/assets/cb81504a-37e5-4a32-bb1d-7abb36579138)
Perhaps this goes in the "Value Resolution" section.

Reviewer: Want to use that exact screenshot?  I'm thinking maybe it's a bit wide and maybe we could put the 2nd table below the first?

Reviewer 3: I was hoping someone would draw a more artistic version. 😄  If the tables are vertically positioned, then maybe the vertical and horizontal axis should also be reversed.

Reviewer 3: I think we could also simply remove the diagram, and rely solely on the text, if we don't want to redraw it per the proposal above.

Reviewer 2: @mel-miller is this something you can help with?

Reviewer 4: @rachelwhitton — We probably shouldn't have this image since it is not accessible (unless the alt text can reference a text-based description of this or the article contains this same thing in text and we can mark this as decorative). 

If you want to keep an image but have a new one drawn you can reach out to creative services (RJ Bravo) to request it.

</dd>

<dt>Value Resolution</dt>

<dd>

> 1. Organization values are resolved first. They form the base value for the key-pair's value resolution.

Reviewer 3: ```suggestion
1. Organization values have the lowest priority. They form the base value that is used when there is no more specific value provided for the site or environment.
```

> 3. Site values are resolved second and secret values set on the site level will override secret values set for the organization. To return the secret to it's organization value, simply delete the site value.

Reviewer 3: ```suggestion
3. Site values will replace the organization values when present. To return the secret to it's organization value, simply delete the site value.
```

> 4. Environmental overrides are resolved finally and if the override exists, it will become the value provided to the calling function.

Reviewer 3: ```suggestion
4. Environmental overrides have the highest priority. If the override exists, it will become the value provided to the calling function.
```

> 5. Each secret's value can be no larger than 16k (16384 Bytes)

Reviewer 3: ```suggestion
Each secret's value can be no larger than 16k (16384 Bytes)
```
